yasa/evaluation.py

```diff
@@ -115,123 +109,403 @@ class EpochByEpochEvaluation:
+        assert all(
+            h1.scorer == h2.scorer for h1, h2 in zip(refr_hyps[:-1], refr_hyps[1:])
+        ), "all `refr_hyps` must have the same scorer"
+        assert all(
+            h1.scorer == h2.scorer for h1, h2 in zip(test_hyps[:-1], test_hyps[1:])
+        ), "all `test_hyps` must have the same scorer"
+        assert all(
+            h1.scorer != h2.scorer for h1, h2 in zip(refr_hyps, test_hyps)
+        ), "each `refr_hyps` and `test_hyps` pair must have unique scorers"
+        assert all(
+            h1.n_epochs == h2.n_epochs for h1, h2 in zip(refr_hyps, test_hyps)
+        ), "each `refr_hyps` and `test_hyps` pair must have the same n_epochs"
+        ## Q: Could use set() for those above.
+        ##    Or set scorer as the first available and check all equal.
+
+        # Convert to dictionaries with sleep_ids and hypnograms
+        refr_hyps = {s: h for s, h in zip(sleep_ids, refr_hyps)}
+        test_hyps = {s: h for s, h in zip(sleep_ids, test_hyps)}
+
+        # Merge all hypnograms into a single MultiIndexed dataframe
+        refr = pd.concat(
+            pd.concat({s: h.as_int()}, names=["sleep_id"]) for s, h in refr_hyps.items()
+        )
+        test = pd.concat(
+            pd.concat({s: h.as_int()}, names=["sleep_id"]) for s, h in test_hyps.items()
+        )
+        data = pd.concat([refr, test], axis=1)
+
+        ########################################################################
+        # INDIVIDUAL-LEVEL AGREEMENT
+        ########################################################################
+
+        # Get individual-level averaged/weighted agreement scores
+        indiv_agree_avg = data.groupby(level=0).apply(self.multi_scorer).apply(pd.Series)
+        ## Q: Check speed against pd.DataFrame({s: multscore(hyps[s], hyps[s]) for s in subjects})
+
+        # Get individual-level one-vs-rest/un-weighted agreement scores
+        # Labels ensures the order of returned scores is known
+        # It also can be used to remove unused labels, but that will be taken care of later anyways
+        # skm_labels = [l for l in refr_hyps[sleep_ids[0]].hypno.cat.categories if l in data.values]
+        # skm will return an array of results, so mapping must be linear without skips
+        ## Q: Another option is to get Series.cat.codes for ints and use cat.categories for mapping
+        skm_labels = np.unique(data).tolist()
+        skm_mapping = {i: l for i, l in enumerate(skm_labels)}  # skm integers to YASA integers
+        mapping_int = refr_hyps[sleep_ids[0]].mapping_int.copy()  # YASA integers to YASA strings
+        # labels = refr_hyps[sleep_ids[0]].labels.copy()  # To preserve YASA ordering
+        # labels = [v for k, v in mapping_int.items() if k in skm_labels]  # To preserve YASA ordering
+        prfs_wrapper = lambda df: skm.precision_recall_fscore_support(
+            *df.values.T, labels=skm_labels, average=None, zero_division=0
+        )
+        indiv_agree_ovr = (
```

Comment:
🤯 Look at that BEAST! This should be the final exam to achieve the Pandas grandmaster title. 

I have actually never used `.explode` or `.pipe`. Will need to run this line by line to make sure I understand

Reply:
I really dislike the way that black formatting handles pandas chaining. There is no exception for pandas, so it forces a new line for every method, whereas I prefer to group each line based on the goal.

Preferred:
```python
df = (data
    # Get precision, recall, f1, and support for each individual sleep session
    .groupby(level=0).apply(prfs_wrapper)
    # Unpack arrays
    .explode().apply(pd.Series)
    # Add metric labels and prepend to index, creating MultiIndex
    .assign(metric=["precision", "recall", "f1", "support"] * len(refr_hyps))
    .set_index("metric", append=True)
    # Convert stage column names to string labels
    ...
```

Black:
```python
df = (data
    # Get precision, recall, f1, and support for each individual sleep session
    .groupby(level=0)
    .apply(prfs_wrapper)
    # Unpack arrays
    .explode()
    .apply(pd.Series)
    # Add metric labels and prepend to index, creating MultiIndex
    .assign(metric=["precision", "recall", "f1", "support"] * len(refr_hyps))
    .set_index("metric", append=True)
    # Convert stage column names to string labels
    ...
```

Are you okay with that long Black structure? Alternatively I could switch them to a series of variables to reduce line count:
```python
# Get precision, recall, f1, and support for each individual sleep session
df= data.groupby(level=0).apply(prfs_wrapper).explode().apply(pd.Series)

# Add metric labels and prepend to index, creating MultiIndex
df = df.assign(metric=["precision", "recall", "f1", "support"] * len(refr_hyps)).set_index("metric", append=True)

# Convert stage column names to string labels
...
```

Reply:
Ah shoot yeah I see what you mean. I'm fine with the current chain structure 👍